README.md
#Cartify
**Version: 2.1.1**

	A Shopping Cart Bundle for Laravel, based on the Cart library from CodeIgniter.

##Installation
**Artisan**

	php artisan bundle:install cartify

**Git Submodule**

	git submodule add git@github.com:bruno-g/cartify.git bundles/cartify

**Modify application/bundles.php**
```php
return array(
	'cartify' => array(
		'handles' => 'cartify', // this is mainly for the examples
		'auto'    => true
	),
);
```

**Publish Assets**

	php artisan bundle:publish

Now visit http://yoursite.com/index.php/cartify and you should see the example products page.



##Adding items to the Cart

####Adding a single item to the Cart
To add an item to the shopping cart, simply pass an array with the product information to the Cartify::cart()->insert() method, as shown below:
```php
$item = array(
	'id'      => 'sku_123ABC',
	'qty'     => 1,
	'price'   => 39.95,
	'name'    => 'T-Shirt',
	'options' => array(
		'size'  => 'L',
		'color' => 'Red'
	)
);

Cartify::cart()->insert($item);
```

####Adding multiple items to the Cart
By using a multi-dimensional array, as shown below, it is possible to add multiple items to the cart in one action.
This is useful in cases where you wish to allow people to select from among several items on the same page.
```php
$items = array(
	array(
		'id'      => 'sku_123ABC',
		'qty'     => 1,
		'price'   => 39.95,
		'name'    => 'T-Shirt',
		'options' => array(
			'size'  => 'L',
			'color' => 'Red'
		)
	),
	array(
		'id'    => 'sku_567ZYX',
		'qty'   => 1,
		'price' => 9.95,
		'name'  => 'Coffee Mug'
	),
	array(
		'id'    => 'sku_965QRS',
		'qty'   => 1,
		'price' => 29.95,
		'name'  => 'Shot Glass'
	)
);

Cartify::cart()->insert($items);
```

####Important:
	The first four array indexes above (id, qty, price, and name) are required.
	If you omit any of them the data will not be saved to the cart.
	The fifth index (options) is optional.
	It is intended to be used in cases where your product has options associated with it.
	Use an array for options, as shown above.

**The five reserved indexes are:**
- **id** - Each product in your store must have a unique identifier. Typically this will be an "sku" or other such identifier.
- **qty** - The quantity being purchased.
- **price** - The price of the item.
- **name** - The name of the item.
- **options** - Any additional attributes that are needed to identify the product. These must be passed via an array.

In addition to the five indexes above, there are two reserved words: **rowid** and **subtotal**. These are used internally by the Cart class, so please do NOT use those words as index names when inserting data into the cart.

Your array may contain additional data. Anything you include in your array will be stored in the session. However, it is best to standardize your data among all your products in order to make displaying the information in a table easier.

The insert() method will return the $rowid if you successfully insert a single item.



##Updating Cart items
To update the information in your cart, you must pass an array containing the Row ID and quantity to the Cartify::cart()->update() method:

####Note:
If the quantity is set to zero, the item will be removed from the cart!

####Updating a single item:
```php
$item = array(
	'rowid'   => 'b99ccdf16028f015540f341130b6d8ec',
	'qty'     => 3,
	'options' => array(
		'size'  => 'M',
		'color' => 'Red'
	)
);

Cartify::cart()->update($item);
```

####Updating multiple items:
```php
$items = array(
	array(
		'rowid'   => 'b99ccdf16028f015540f341130b6d8ec',
		'qty'     => 3,
		'options' => array(
			'size'  => 'M',
			'color' => 'Red'
		)
	),
	array(
		'rowid' => 'xw82g9q3r495893iajdh473990rikw23',
		'qty'   => 4
	),
	array(
		'rowid'   => 'fh4kdkkkaoe30njgoe92rkdkkobec333',
		'qty'     => 2,
		'options' => array(
			'color' => 'Yellow'
		)
	)
);

Cartify::cart()->update($items);
```

####What is a Row ID?
The row ID is a unique identifier that is generated by the cart code when an item is added to the cart. The reason a unique ID is created is so that identical products with different options can be managed by the cart.

For example, let's say someone buys two identical t-shirts (same product ID), but in different sizes. The product ID (and other attributes) will be identical for both sizes because it's the same shirt. The only difference will be the size. The cart must therefore have a means of identifying this difference so that the two sizes of shirts can be managed independently. It does so by creating a unique "row ID" based on the product ID and any options associated with it.

In nearly all cases, updating the cart will be something the user does via the "view cart" page, so as a developer, it is unlikely that you will ever have to concern yourself with the "row ID", other then making sure your "view cart" page contains this information in a hidden form field, and making sure it gets passed to the update function when the update form is submitted. Please examine the construction of the "view cart" page below for more information.



##Removing items from the Cart
To remove an item from your cart, you must pass the Row ID to the Cartify::cart()->remove() method:
```php
Cartify::cart()->remove('fh4kdkkkaoe30njgoe92rkdkkobec333');
```


##Displaying the Cart
To display the cart you will create a view file with code similar to the one shown in cart.blade.php view file.



## Function Reference
```php
Cartify::cart()->item(rowid)
```
Returns all the information about an item from the shopping cart.

```php
Cartify::cart()->insert();
```
Permits you to add items to the shopping cart.

```php
Cartify::cart()->update();
```
Permits you to update items in the shopping cart.

```php
Cartify::cart()->remove(rowid);
```
Permits you to remove an item from the shopping cart.

```php
Cartify::cart()->total();
```
Displays the total amount in the cart.

```php
Cartify::cart()->total_items();
```
Displays the total number of items in the cart.

```php
Cartify::cart()->contents();
```
Returns an array containing everything in the cart.

```php
Cartify::cart()->has_options(rowid);
```
Returns TRUE (boolean) if a particular row in the cart contains options. This function is designed to be used in a loop with Cartify::cart()->contents(), since you must pass the rowid to this function, as shown in the Displaying the Cart example above.

```php
Cartify::cart()->item_options(rowid);
```
Returns an array of options for a particular item. This function is designed to be used in a loop with Cartify::cart()->contents(), since you must pass the rowid to this function, as shown in the Displaying the Cart example above.

```php
Cartify::cart()->destroy();
```
Permits you to destroy the cart. This function will likely be called when you are finished processing the customer's order.


## Multiple instances of the Cart
It is possible to have multiple instances of the cart class.

You just need to pass in the name of the cart like so:
```php
Cartify::cart('my_other_cart')->insert($item);
```

It is included an Whishlist method to help you, and you can even create multiple wishlist's aswell, like so:
```php
Cartify::wishlist()->insert($item);
Cartify::wishlist('my_other_wishlist')->insert($item);
```



##Exceptions

Some methods on the Cart library throws Exceptions, below you have a list of all the Exceptions.

	Cartify\CartException

This Exception is Thrown when an unexpected error occurred.

	Cartify\CartInvalidDataException

This Exception is Thrown when an array is not passed when inserting or updating cart items.

	Cartify\CartItemNotFoundException

This Exception is Thrown when an item is not found in the shopping cart.

	Cartify\CartRequiredIndexException

This Exception is Thrown when a required index is missing from the array.

	Cartify\CartInvalidItemRowIdException

This Exception is Thrown when an Item Row ID is invalid.

	Cartify\CartInvalidItemNameException

This Exception is Thrown when an Item Name is invalid.

	Cartify\CartInvalidItemQuantityException

This Exception is Thrown when the quantity is invalid.

	Cartify\CartInvalidItemPriceException

This Exception is Thrown when an Item Price is invalid.

	To better understand on how to use the Exceptions, you can check the example controllers.



## Credits

	CodeIgniter for writing the Cart Library.
	Twitter for the awesome Bootstrap framework.


CHANGELOG.md
**Version: 2.1.1**

	* Cart class code improvements.
	* Adding an item that already exists on the shopping cart, the item quantity on the shopping cart is incremented.
	* Some views updates.
	* Using more Laravel helper functions.

**Version: 2.1.0**

	* Added a new item() method to the cart class.
	* Now almost every method from the cart library throws an Exception.
	* Renamed the cart add() method to insert()
	* Added an helpers file.
	* format_number() is no longer a method in the cart library, it is now a function inside the helpers file.
	* Code improvements.
	* Updated the examples.
	* Added the public folder, which contains the needed assets for the examples.
	* Updated to Twitter Bootstrap 2.2.1
	* Item names can now have accents.

**Version: 2.0.0**

    * Changed bundle name from Lara-Cart to Cartify
    * Completely refactored!
    * It is now possible to have multiple instances of the cart.
    * Added in a Wishlist method.

**Version: 1.0**

    * Everything is new.


models/products.php
<?php
/**
 * --------------------------------------------------------------------------
 * Cartify
 * --------------------------------------------------------------------------
 *
 * Cartify, a shopping cart bundle for use with the Laravel Framework.
 *
 * @package  Cartify
 * @version  2.1.1
 * @author   Bruno Gaspar <brunofgaspar1@gmail.com>
 * @link     https://github.com/bruno-g/cartify
 */

namespace Cartify\Models;


#############################################################
############ This file is for the examples only! ############
#############################################################


/**
 * Products, static model.
 */
class Products
{
	/**
	 * Returns a list of all products.
	 *
	 * @access   public
	 * @return   array
	 */
	public static function all()
	{
		// Declare our static products, this is just for the example !!!
		//
		return array(
			// Product 1
			//
			'sku_123ABC' => array(
				'id'      => 'sku_123ABC',
				'price'   => 39.95,
				'name'    => 'T-Shirt',
				'image'   => 'tshirt.jpg',
				'options' => array(
					'Size'  => array(
						's' => 'S',
						'm' => 'M',
						'l' => 'L'
					),
					'Color' => array(
						'red'    => 'Red',
						'blue'   => 'Blue',
						'yellow' => 'Yellow',
						'white'  => 'White'
					),
					'Style' => array(
						'unisex' => 'Unisex',
						'womens' => 'Womens'
					)
				)
			),

			// Product 2
			//
			'sku_567ZYX' => array(
				'id'      => 'sku_567ZYX',
				'price'   => 9.95,
				'name'    => 'Coffee Mug',
				'image'   => 'coffee_mug.jpg',
				'options' => array(
					'Design' => array(
						'design 1'=> 'Design 1',
						'design xpt0' => 'Design XpT0'
					)
				)
			),

			// Product 3
			//
			'sku_965QRS' => array(
				'id'      => 'sku_965QRS',
				'price'   => 29.95,
				'name'    => 'Shot Glass',
				'image'   => 'shot_glass.jpg'
			)
		);
	}

	/**
	 * Returns the options of a product.
	 *
	 * @access   public
	 * @param    string
	 * @return   array
	 */
	public static function get_options($item_id)
	{
		// Get the list of products.
		//
		$products = static::all();

		// Return the product options.
		//
		return array_get($products, $item_id . '.options', array());
	}

	/**
	 * Returns the information about a product.
	 *
	 * @access   public
	 * @param    string
	 * @return   mixed
	 */
	public static function find($product_slug = null)
	{
		// Get the list of products.
		//
		$products = static::all();

		// Loop through the products.
		//
		foreach ($products as $product)
		{
			// Check if this the product we are looking for.
			//
			if (\Str::slug($product['name']) === $product_slug)
			{
				// Return the product information.
				//
				return $product;
			}
		}

		// Product was not found.
		//
		return false;
	}
}
